kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,215 @@
+
+```yaml
+  system-critical: 10000
+             high:  9000
+              low:  8000
+low-non-preempted:  8000 # with minimumPreemptablePriority=10000
+                        # (i.e. high can't preempt this, but system-critical can preempt this)
+```
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. As described above, job would need check-pointing. However, most machine learning jobs are iteration-based process, called _epochs_. So, if the cluster is high load and preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and computational resources consumed by the job came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+type PreemptionToleration struct {
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.
+  // Nil value means forever, which is not preempted forever (default)
+  // zero and negative value means immediate, which is preempted immediately. i.e. no toleration at all.
+  // This value affects scheduled pods only (no effect on nominated pods).
+  TolerationSeconds *int64 `json:"tolerationSeconds,omitempty"`
+
+  // more policy will arise in the future.
+}
+```
+
+Users specifies the preemption toleration policy as an annotation with the key `preemption-toleration.scheduling.sigs.k8s.io/<property_name>` in `PriorityClass` as below:
+
+```yaml
+kind: PriorityClass
+metadata:
+  name: toleration-policy-sample
+  annotation:
+    # this key is needed to enable preemption toleration policy for distinguishing 
```

Comment:
Hmm... this is another shortage of annotations (vs. CRD). It's ok to start with annotations as an alpha feature, but once we graduate it to beta, do you have any plan to turn to CRD? - which may better manage the specs.